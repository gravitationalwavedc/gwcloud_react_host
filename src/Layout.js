import {Grid, Menu} from "semantic-ui-react";
import React from "react";
import Link from 'found/lib/Link';

function Layout(props) {
    console.log(props)
    return (
        <div>
            <Menu fixed={'top'}>
                {/*<Container>*/}
                <Menu.Item as='a' header>
                    GW-Cloud
                </Menu.Item>
                <Menu.Item as='div'><Link to="/" activeClassName="selected" exact>Home</Link></Menu.Item>
                {/*</Container>*/}
            </Menu>
            <Grid columns={1}>
                <Grid.Column>
                    {props.children}
                </Grid.Column>
            </Grid>
        </div>
    )
}

export default Layout;