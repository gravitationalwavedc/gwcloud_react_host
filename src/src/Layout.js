import React from "react";
import {createFragmentContainer} from "react-relay";
import Menu from "./components/Menu";
import HarnessApi from "./HarnessApi";

const noMenuURLs = [
  "/auth/"
];

const setUser = (user) => {
  // Update the current user
  HarnessApi.currentUser = user;
  return user;
};

const Layout = ({gwclouduser, children, match}) => {
    const user = setUser(gwclouduser);
    const name = user ? `${user.firstName} ${user.lastName}` : null;
    const showMenu= !noMenuURLs.includes(match.location.pathname);
    return (
      <React.Fragment>
        <header>
          {showMenu && <Menu name={name}/>}
        </header>
        <main style={{paddingTop: '56px'}}>
          {children}
        </main>
      </React.Fragment>
    )
}

export default createFragmentContainer(Layout, {
    gwclouduser: graphql`
        fragment Layout_gwclouduser on UserDetails {
          userId
          username
          firstName
          lastName
        }
    `
});
