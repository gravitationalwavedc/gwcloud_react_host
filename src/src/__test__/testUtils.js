export function mockLocationHostName(hostname) {
    // Wild hax from github: https://github.com/facebook/jest/issues/890#issuecomment-456804379
    delete global.window.location;
    global.window = Object.create(window);
    global.window.location = {
        hostname: hostname
    };
}