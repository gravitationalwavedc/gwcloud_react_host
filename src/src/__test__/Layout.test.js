import React from "react";
import {expect, jest} from "@jest/globals";
import TestRenderer from 'react-test-renderer';
import Layout from "../Layout";
import {createMockEnvironment, MockPayloadGenerator} from "relay-test-utils";
import {QueryRenderer} from 'react-relay';
import {mockLocationHostName} from "./testUtils";

jest.mock('found', () => ({
    Link: component => <a {...component}>{component.children}</a>
}))

test('Layout renders login if user is not set', () => {
    mockLocationHostName("gwlab.org.au");

    const environment = createMockEnvironment();
    const MyTestRenderer = () => (
        <QueryRenderer
            environment={environment}
            query={graphql`
                query LayoutUserNotSetTestQuery @relay_test_operation {
                    gwclouduser {
                        ...Layout_gwclouduser
                    }
                }
              `}
            variables={{}}
            render={({error, props}) => {
                if (props) {
                    return <Layout {...props} match={{location: {pathname: '/'}}}/>;
                } else if (error) {
                    return error.message;
                }
                return 'Loading...';
            }}
        />
    );

    const renderer = TestRenderer.create(<MyTestRenderer/>);
    environment.mock.resolveMostRecentOperation(operation =>
        MockPayloadGenerator.generate(operation, {
            gwclouduser() {
                return {
                    userId: null,
                    username: null,
                    firstName: null,
                    lastName: null
                }
            }
        }),
    );

    // todo: Check that login is in the document
    expect(renderer).toMatchSnapshot();
});

test('Layout renders login if user is set', () => {
    mockLocationHostName("gwlab.org.au");

    const environment = createMockEnvironment();
    const MyTestRenderer = () => (
        <QueryRenderer
            environment={environment}
            query={graphql`
                query LayoutUserSetTestQuery @relay_test_operation {
                    gwclouduser {
                        ...Layout_gwclouduser
                    }
                }
              `}
            variables={{}}
            render={({error, props}) => {
                if (props) {
                  return <Layout {...props} match={{location: {pathname: '/'}}}/>;
                } else if (error) {
                    return error.message;
                }
                return 'Loading...';
            }}
        />
    );

    const renderer = TestRenderer.create(<MyTestRenderer/>);
    environment.mock.resolveMostRecentOperation(operation =>
        MockPayloadGenerator.generate(operation, {
            gwclouduser() {
                return {
                    userId: 4,
                    username: "billnye",
                    firstName: "Bill",
                    lastName: "Nye"
                }
            }
        }),
    );

    // todo: Check that the user name and logout is in the document
    expect(renderer).toMatchSnapshot();
});
