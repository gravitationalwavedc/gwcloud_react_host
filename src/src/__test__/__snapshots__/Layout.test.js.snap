// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout renders login if user is not set 1`] = `
Array [
  <header>
    <nav
      className="navbar navbar-expand navbar-light fixed-top"
    >
      <span
        className="mr-auto navbar-brand"
      >
        <a
          exact={true}
          to="/"
        >
          GW Cloud
        </a>
      </span>
      <div
        className="navbar-nav"
        onKeyDown={[Function]}
      >
        <span
          className="justify-content-end navbar-text"
        >
          &lt;mock-value-for-field-"firstName"&gt; &lt;mock-value-for-field-"lastName"&gt;
        </span>
        <a
          className="nav-link"
          data-rb-event-key={null}
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
        >
          Logout
        </a>
      </div>
    </nav>
  </header>,
  <main
    style={
      Object {
        "paddingTop": "56px",
      }
    }
  />,
]
`;

exports[`Layout renders login if user is set 1`] = `
Array [
  <header>
    <nav
      className="navbar navbar-expand navbar-light fixed-top"
    >
      <span
        className="mr-auto navbar-brand"
      >
        <a
          exact={true}
          to="/"
        >
          GW Cloud
        </a>
      </span>
      <div
        className="navbar-nav"
        onKeyDown={[Function]}
      >
        <span
          className="justify-content-end navbar-text"
        >
          &lt;mock-value-for-field-"firstName"&gt; &lt;mock-value-for-field-"lastName"&gt;
        </span>
        <a
          className="nav-link"
          data-rb-event-key={null}
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
        >
          Logout
        </a>
      </div>
    </nav>
  </header>,
  <main
    style={
      Object {
        "paddingTop": "56px",
      }
    }
  />,
]
`;
